README.md
## Idle Orbs 3


The third installment of the Idle Orbs Series™!

Gh Pages Link: https://jacoby-y.github.io/idle-orbs-3/


#### Front Burner
- Traits gui: select, see info, and scrap traits
- Auto bounce
- Render modes
- Scrap orb and remove it from physics
- Fix double click scrapping on orbs


#### Back Burner
- Prestige and paradigm shift

changelog.md
## v0.1.1
- Code restructuring
- Fixed small bugs

## v0.1
- A very well functioning first commit

src/lib/draw.js
import { ctx as _ctx } from "../stores.js";
/** @type {CanvasRenderingContext2D} */
let ctx; 
_ctx.subscribe(v => ctx = v);

export default {
	circle(x,y, r, color, fill=true, line_width=1) {
		ctx.beginPath();
		ctx.arc(x,y,r,0,2*Math.PI);
		if (fill) {
			ctx.fillStyle = color;
			ctx.fill();
		}
		else { 
			ctx.lineWidth = line_width; 
			ctx.strokeStyle = color; 
			ctx.stroke(); 
		}
	},
	rect(x,y,w,h, color, fill=true, line_width=1) {
		if (fill) {
			ctx.fillStyle = color;
			ctx.fillRect(x,y,w,h);
		} else {
			ctx.strokeStyle = color;
			ctx.lineWidth = line_width;
			ctx.strokeRect(x,y,w,h);
		}
	},
	line1(x1, y1, x2, y2, color, width) {
		ctx.strokeStyle = color;
		ctx.lineWidth = width;
		ctx.beginPath();
		ctx.moveTo(x1, y1);
		ctx.lineTo(x2, y2);
		ctx.stroke();
	},
	lineAtAngle(x,y, ang, len, color, line_width=1) {
		ctx.strokeStyle = color;
		ctx.lineWidth = line_width;
		ctx.beginPath();
		ctx.moveTo(x,y);
		ctx.lineTo(x+Math.cos(ang)*len, y+Math.sin(ang)*len);
		ctx.stroke();
	},
	clear(x,y,w,h) {
		ctx.clearRect(x,y,w,h);
	},
	transform(x=0, y=0, rot=0) {
		ctx.translate(x,y);
		ctx.rotate(rot);
	},
	resetTransform() { ctx.setTransform(1, 0, 0, 1, 0, 0); },

};

src/lib/orbs.js
import { size, cash, render_mode, orbs as _orbs, rarity_key, auto_bounce, coll_rad, calcTrait, tps } from "../stores.js";
import draw from "../lib/draw.js";
import { abs, distance1, distance2, fixAng, numDiff, round, runNTimes } from "../lib/funcs.js";
import { get } from "svelte/store";

export const OrbM = (()=>{
	const collect = (orb)=>{
		cash.update(v => v + (calcTrait.value(orb)/(tps/2)));
	};

	const phys = {
		l: [],
		new(orb) { this.l.push(orb) },
		step(i) {
			const orb = this.l[i];
			orb.x += orb.vx;
			orb.y += orb.vy;

			if (orb.x-10 < 0) {
				orb.x = 10;
				orb.vx = abs(orb.vx);
			} else if (orb.y-10 < 0) {
				orb.y = 10;
				orb.vy = abs(orb.vy);
			}
			if (orb.x+10 > size.w) {
				orb.x = size.w-10;
				orb.vx = abs(orb.vx) * -1;
			} else if (orb.y > size.h-10) {
				orb.y = size.h-10;
				orb.vy = abs(orb.vy) * -1;
			}
		}
	};

	let orbs = [];
	_orbs.subscribe(v => orbs = v);
	
	return {
		update() {
			for (let i = 0; i < orbs.length; i++) {
				const orb = orbs[i];
				if (phys.l[i] == undefined) {
					phys.new({
						x: size.w*Math.random(), y: size.h*Math.random(),
						vx: 10, vy: 10,
					});
				}
				const { x, y } = phys.l[i];
				const ph = phys.l[i];
				
				ph.vy += 0.8 - (calcTrait.lessGravity(orb)*0.1);
				ph.vx *= Math.min(1, 0.99+(0.3*calcTrait.lessDrag(orb)));
				ph.vy *= Math.min(1, 0.99+(0.3*calcTrait.lessDrag(orb)));

				const coll_dist = distance1(ph.x, ph.y, size.w/2, size.h/3);
				if (coll_dist < get(coll_rad)) {
					collect(orb);
					draw.line1(size.w/2, size.h/3, x, y, "#99990066", 4);
				}

				const magnetic = calcTrait.magnetic(orb);
				if (magnetic > 0 && coll_dist > 50) {
					const coll_ang = Math.atan2(size.h/3-ph.y, size.w/2-ph.x);
					// draw.lineAtAngle(ph.x, ph.y, coll_ang, 100, "lime");
					ph.vx += Math.cos(coll_ang)/3 * magnetic;
					ph.vy += Math.sin(coll_ang)/3 * magnetic;
				}
				
				draw.circle(x,y, 10, rarity_key[orb.rarity].color);
				phys.step(i);
			}
		},
		bounce(auto=false) {
			for (let i = 0; i < phys.l.length; i++) {
				const ph = phys.l[i];
				if (ph.y < size.h - get(auto_bounce).size-20) continue;
				ph.vy -= 25 + 10*Math.random();
				if (auto) ph.vx += (size.w/2 - ph.x)/70;
				else if (this.mouse_x >= 0) ph.vx += (this.mouse_x - ph.x)/70;
				ph.vx += -0.5 + Math.random()
			}
		},
		mouse_x: -1,
		info: orbs, 
		get phys() { return phys; },
	};
})();

src/lib/funcs.js
export const exp = (pow=1)=> (10 ** pow)
export const round = (num, pow=1)=> Math.round(num * exp(pow))/exp(pow);
export const floor = (num, pow=1)=> Math.floor(num * exp(pow))/exp(pow);
export const ceil = (num, pow=1)=> Math.ceil(num * exp(pow))/exp(pow);
export const toExp = (num, place)=>{
  let pow = Math.floor(Math.log10(num));
	place = 10**place;
  return (Math.floor(num/(10**pow)*place)/place + `e${pow}`);
}
export const sci = (num, place=1)=> num >= 1000 ? toExp(num, place).replace("+", "") : round(num);
export const fixBigNum = (num, place)=>{
  const l = Math.floor(Math.log10(num));
  return Math.round(num / ((10**l)/(10**place)))*((10**l)/(10**place));
}
export const formatSeconds = (secs, short=true)=>{
  let neg = secs < 0;
  if (neg) secs = Math.abs(secs);
  if (secs < 60) return `${neg ? '-' : ''}${round(secs, 0)}${short ? "s" : " Seconds"}`;
  secs /= 60;
  if (secs < 60) return `${neg ? '-' : ''}${ceil(secs)}${short ? "m" : " Minutes"}`;
  secs /= 60;
  if (secs < 60) return `${neg ? '-' : ''}${ceil(secs)}${short ? "h" : " Hours"}`;
  secs /= 24;
  if (secs < 24) return `${neg ? '-' : ''}${ceil(secs)}${short ? "d" : " Days"}`;
  secs /= 365;
  return `${neg ? '-' : ''}${ceil(secs)}${short ? "y" : " Years"}`;
}
export const runNTimes = (n, callback)=>{
  Array.from(Array(n)).forEach(callback);
}
export const floorRound = (num, place)=>{
  const pow = (Math.pow(10, place));
  return Math.floor(num * pow) / pow;
}
const num_shorts = ['', 'K', 'M', 'B', 'T', 'Qa', 'Qi', 'Sx', 'Sp', 'O', 'N', 'D', 'UD', 'DD', 'TD', 'QuD', 'QiD', 'SxD', 'SpD', 'OD', 'ND', 'V', 'UV', 'DV', 'TV', 'QaV', 'QiV', 'SxV', 'SpV', 'OV', 'NV', 'T', 'UT', 'DT', 'TT', 'QaT', 'QiT', 'SxT', 'SpT', 'OT', 'NT'];
export const fNum = (num, round_to=1, i=0, past_thresh=false)=>{
    const div = num / 1000;
    const thresh = (i >= num_shorts.length);
    if (div < 1 || thresh) { 
      if (thresh) return (floorRound(num, round_to) + num_shorts[num_shorts.length-1]);
      else return (i == 0) ? (floorRound(num, round_to)) : (floorRound(num, round_to) + num_shorts[i]);
    }
    return fNum(div, round_to, i+1, thresh);
}

/** c: cash | x: starting about | m: increase */
export const multComp = (c, x, m)=>{
  // c = x * (m^n);
  // c/x = m^n
  // log(c/x)/log(m) = n
  const pow = Math.floor(Math.log(c/x)/Math.log(m));
  const total = Math.floor(x * (m ** pow))

  return {
    pow, 
    total,
  }
}
/** c: starting cost | i: increments | x: increase */
export const addComp = (c, i, x)=>{
  // 50 * 5 + 10 * m;
  const m = (i-1)/2 * i;
  return c * i + x * m;
}
// console.log(addComp(50, 6, 10));

let run_n_save = null;
export const runN = (func=()=>{}, n)=>{
  if (run_n_save == func) return;
  run_n_save = func;
  for (let i = 0; i < n; i++) func();
  run_n_save = null;
}

export const spendCashAdd = (cash, cost, incr)=>{
  let i = 0;
  if (incr == 0) {
    i = Math.floor(cash/cost);
    cash -= i*cost;
  } else {
    for ([]; cost <= cash; (cost += incr, i++)) {
      cash -= cost;
    }
  }
  return {
    cash,
    cost,
    i,
  }
}
export const spendCashMult = (cash, cost, incr)=>{
  let i = 0;
  if (incr == 0) {
    i = Math.floor(cash/cost);
    cash -= i*cost;
  } else {
    for ([]; cash > cost; (cost *= incr, i++)) {
      cash -= cost;
      console.log(cash);
    }
  }
  return {
    cash,
    cost,
    i,
  }
}

export const distance1 = (x1,y1, x2,y2)=>{
  return Math.sqrt((x2-x1) * (x2-x1) + (y2-y1) * (y2-y1));
}
export const distance2 = (pos1, pos2)=>{
  return Math.sqrt((pos2.x-pos1.x) * (pos2.x-pos1.x) + (pos2.y-pos1.y) * (pos2.y-pos1.y));
}

export const fixAng = (ang)=> (ang + Math.PI) % (Math.PI * 2);
export const abs = Math.abs;
export const numDiff = (num1, num2)=>{
  num1 = abs(num1); num2 = abs(num2);
  const max = Math.max(num1, num2);
  const min = Math.min(num1, num2);
  return max - min;
}

export const randRange = (min, max)=> Math.floor(Math.random() * ((max+1) - min) + min);

export const toTitle = (str)=>{
  if (!str) return;
  /** @type String[] */
  const spl = str.split(" ");
  for (let i = 0; i < spl.length; i++) {
    spl[i] = spl[i][0].toUpperCase() + spl[i].slice(1);
  }
  return spl.join(" ");
}

src/stores.js
import { get, writable } from "svelte/store";
import { randRange } from "./lib/funcs.js";

//#region | Cash / Dust / Timer / Tps
export const dust = writable(0);
let deci = 0;
export const cash = writable(0);
cash.subscribe((v)=>{
	if (v < 0) { cash.set(0); return; }
	if (Math.floor(v) == v) return;
	deci += v - Math.floor(v);
	if (deci >= 1) {
		v += Math.floor(deci);
		deci -= Math.floor(deci);
		cash.set(Math.floor(v));
	} else {
		cash.set(Math.floor(v));
	}
});
export const timer = writable(0);
export const tps = 45;

const run_timer = ()=> timer.update(v => (v+1)%tps);
let timer_interval = setInterval(run_timer, 1000/tps);
//#endregion

//#region | Canvas
/** @type {CanvasRenderingContext2D}*/ let _ctx;
export const ctx = writable(_ctx);

export const size = { w: 1600, h: 900 };
//#endregion

//#region | Settings
export const render_mode = writable(0);
//#endregion

//#region | Orb Stuff
export const orbs = writable([
	{ rarity: "common", value: 1, sockets: [] },
	// { rarity: "uncommon", value: 2, sockets: ["", ""] },
	// { rarity: "rare", value: 3, sockets: ["", "", "", ] },
	// { rarity: "epic", value: 4, sockets: ["", "", "", "", ] },
	// { rarity: "legendary", value: 5, sockets: ["", "", "", "", "", ""] },
	// { rarity: "magic", value: 6, sockets: ["", "", "", "", ""] },
]);
export const auto_bounce = writable({
	prog: 0,
	// max: 20,
	max: tps*3,
	unlocked: false,
	unlock_cost: 100,
	size: 100,
	speed_cost: 100,
	size_cost: 100,
});
export const collect_radius = writable(100);
export const addNewOrb = ()=>{
	let perc = Math.round(Math.random()*100);
	const [r, obj] = (()=>{
		for (const k in rarity_key) {
			if (!Object.hasOwnProperty.call(rarity_key, k)) continue;
			const v = rarity_key[k];
			perc -= v.chance;
			if (perc <= 0) return [k, v];
		}
	})()
	const new_orb = { rarity: "common", value: 1, sockets: [] };
	new_orb.rarity = r;
	new_orb.value = obj.get_value();
	new_orb.sockets = Array.from(Array(obj.get_sockets())).map(()=> "");

	orbs.update(v => [ ...v,  new_orb]);
}
export const coll_rad = writable(200);
//#endregion

//#region | Traits
export const traits = writable([ ]);
const new_trait = (desc, weight)=>{return {desc, weight} }
export const trait_key = {
	helium: new_trait("Orb is lighter.", 15),
	ice: new_trait("Orb takes less drag", 15),
	midas: new_trait("Orb value increases by $3.", 5),
	magnetic: new_trait("Orb is attracted to the collector.", 1),
};
const total_trait_weight = Object.keys(trait_key).reduce((p, c)=> p+trait_key[c].weight, 0);
export const trait_cost = {
	cash: 50,
	dust: 10,
}
export const addNewTrait = ()=>{
	const rand = randRange(0, total_trait_weight);
	let new_t = "midas";
	Object.keys(trait_key).reduce((p, c)=> {
		const res = p+trait_key[c].weight;
		if (rand >= p && rand < res) new_t = c;
		return res;
	}, 0);
	traits.update(v => [...v, new_t]);
}
export const calcTrait = {
	value(orb) {
		if (!orb) return;
		return orb.value + orb.sockets.reduce((p,c)=> c == "midas" ? p+1 : p, 0) * 3;
	},
	lessDrag(orb) {
		if (!orb) return;
		return orb.sockets.reduce((p,c)=> c == "ice" ? p+1 : p, 0);
	},
	lessGravity(orb) {
		if (!orb) return;
		return orb.sockets.reduce((p,c)=> c == "helium" ? p+1 : p, 0);
	},
	magnetic(orb) {
		if (!orb) return;
		return orb.sockets.reduce((p,c)=> c == "magnetic" ? p+1 : p, 0);
	}
};
//#endregion

//#region | Rarities
export const rarity_key = {
	common: {
		scrap_value: 5,
		chance: 40,
		color: "white",
		get_value: ()=> randRange(1,1),
		get_sockets: ()=> randRange(0, 1),
	}, 
	uncommon: {
		scrap_value: 10,
		chance: 20,
		color: "lime",
		get_value: ()=> randRange(2,5),
		get_sockets: ()=> randRange(1, 2),
	}, 
	rare: {
		scrap_value: 20,
		chance: 15,
		color: "aqua",
		get_value: ()=> randRange(5,10),
		get_sockets: ()=> randRange(2, 3),
	}, 
	epic: {
		scrap_value: 40,
		chance: 10,
		color: "violet",
		get_value: ()=> randRange(10,15),
		get_sockets: ()=> randRange(3, 4),
	}, 
	legendary: {
		scrap_value: 70,
		chance: 5,
		color: "gold",
		get_value: ()=> randRange(15,20),
		get_sockets: ()=> randRange(4, 5),
	}, 
	magic: {
		scrap_value: 50,
		chance: 10,
		color: "pink",
		get_value: ()=> randRange(5,30),
		get_sockets: ()=> randRange(1, 6),
	}
};
//#endregion

window.onbeforeunload = ()=>{
	clearInterval(timer_interval);
}
